libvast/src/system/index.cpp

```diff
@@ -41,24 +47,692 @@
+  }
+}
+
+query_map
+index_state::launch_evaluators(pending_query_map& pqm, expression expr) {
+  query_map result;
+  for (auto& [id, eval] : pqm) {
+    std::vector<caf::actor> xs{self->spawn(evaluator, expr, std::move(eval))};
+    result.emplace(id, std::move(xs));
+  }
+  pqm.clear();
+  return result;
+}
+
+path index_state::index_filename(path basename) const {
+  return basename / dir / "index.bin";
+}
+
+caf::expected<flatbuffers::Offset<fbs::Index>>
+pack(flatbuffers::FlatBufferBuilder& builder, const index_state& state) {
+  auto meta_idx = pack(builder, state.meta_idx);
+  if (!meta_idx)
+    return meta_idx.error();
+  VAST_VERBOSE(state.self, "persisting", state.persisted_partitions.size(),
+               " definitely persisted and ", state.unpersisted.size(),
+               " maybe persisted partitions uuids");
+  std::vector<flatbuffers::Offset<fbs::UUID>> partition_offsets;
+  for (auto uuid : state.persisted_partitions) {
+    auto uuid_fb = pack(builder, uuid);
+    if (!uuid_fb)
+      return uuid_fb.error();
+    partition_offsets.push_back(*uuid_fb);
+  }
+  // We don't know if these will make it to disk before the index and the rest
+  // of the system is shut down (in case of a hard/dirty shutdown), so we just
+  // store everything and throw out the missing partitions when loading the
+  // index.
+  for (auto& kv : state.unpersisted) {
+    auto uuid_fb = pack(builder, kv.first);
+    if (!uuid_fb)
+      return uuid_fb.error();
+    partition_offsets.push_back(*uuid_fb);
+  }
+  auto partitions = builder.CreateVector(partition_offsets);
+  std::vector<flatbuffers::Offset<fbs::LayoutStatistics>> stats_offsets;
+  for (auto& [name, layout_stats] : state.stats.layouts) {
+    auto name_fb = builder.CreateString(name);
+    fbs::LayoutStatisticsBuilder stats_builder(builder);
+    stats_builder.add_name(name_fb);
+    stats_builder.add_count(layout_stats.count);
+    auto offset = stats_builder.Finish();
+    stats_offsets.push_back(offset);
+  }
+  auto stats = builder.CreateVector(stats_offsets);
+  fbs::IndexBuilder index_builder(builder);
+  index_builder.add_version(fbs::Version::Index_v0);
+  index_builder.add_meta_index(*meta_idx);
+  index_builder.add_partitions(partitions);
+  index_builder.add_stats(stats);
+  return index_builder.Finish();
+}
+
+/// Persists the state to disk.
+void index_state::flush_to_disk() {
+  auto builder = new flatbuffers::FlatBufferBuilder();
+  auto index = pack(*builder, *this);
+  if (!index) {
+    VAST_WARNING(self, "couldnt pack index", index.error());
+    return;
+  }
+  builder->Finish(*index, vast::fbs::file_identifier);
+  auto ptr = builder->GetBufferPointer();
+  auto size = builder->GetSize();
+  auto chunk = vast::chunk::make(size, ptr, [=] { delete builder; });
+  self
+    ->request(caf::actor_cast<caf::actor>(filesystem), caf::infinite,
+              atom::write_v, index_filename(), chunk)
+    .then([=](atom::ok) { VAST_DEBUG(self, "finished writing to disk"); },
```

Comment:
```suggestion
    .then([=](atom::ok) { VAST_DEBUG(self, "successfully persisted index state"); },
```

I don't like to use the word "disk" because it assumes a specific hardware setting. Let's go with "persisting" instead.

Reply:
I hope nobody wants to persist indexer state on blockchainfs anytime soon ;) Changed it anyways.